feat(Combobox): add categories feature

akdetrick opened the pull request:
closes #1018 

**blocked by: https://github.com/narmi/design_system/pull/1023**

Adds `Combobox.Category` support, similar to the category feature in `Select`.

## How it works

When categories are passed into `Combobox`, we extract all the children as a flat list, [similar to `Select`](https://github.com/narmi/design_system/blob/a6f155ba42c9d1edb087b326b35b582487be78c1/src/Select/index.js#L129). This allows the downshift `useCombobox` hook to ignore categories and behave the same as a basic Combobox.

Categories are used to nest items inside `details` elements when rendering, [similar to how it works in `Select`](https://github.com/narmi/design_system/blob/a6f155ba42c9d1edb087b326b35b582487be78c1/src/Select/index.js#L282).

### Category rendering rules

A category is only rendered when:
- It has at least one item that has not been filtered out by the typeahead behavior

A category is only opened when:
- User clicks on the category name
- Highlight via mouse or keyboard nav moves onto an item included in the category
- When items are being filtered, so users can see the category names _and_ the suggestions while filtering

## Demo
![Sep-08-2023 17-35-19](https://github.com/narmi/design_system/assets/231252/49553e2b-a4b7-42a0-ac53-4b54274d0218)

akdetrick commented on src/Combobox/index.js:
```diff
@@ -107,29 +173,115 @@ const Combobox = ({
-    triggerOffset: -3,
-    containerOffset: 16,
-  });
+  const { renderLayer, triggerProps, layerProps, triggerBounds, layerSide } =
+    useLayer({
+      isOpen,
+      overflowContainer: true,
+      auto: true,
+      snap: true,
+      placement: "bottom-start",
+      possiblePlacements: ["top-start", "bottom-start"],
+      triggerOffset: -3,
+      containerOffset: 16,
+    });
```
formatting only

akdetrick commented on src/Combobox/index.js:
```diff
@@ -179,58 +333,21 @@ const Combobox = ({
             "list--reset bgColor--white border--right bottom border--left",
             {
-              "nds-combobox-list--active": isOpen,
+              "nds-combobox-list--active": isOpen && displayedItems.length > 0,
```
Fixes existing bug where the popup still rendered when empty

akdetrick commented on src/Combobox/ComboboxItem.js:
```diff
@@ -14,6 +14,12 @@ ComboboxItem.propTypes = {
    */
   value: PropTypes.string.isRequired,
+  /**
+   * Custom typeahead string. By default typeahead uses `value`.
```
```suggestion
   * Custom typeahead string. By default, typeahead uses `value`.
```

akdetrick commented on src/Combobox/index.js:
```diff
@@ -26,6 +28,57 @@ export const isSelectable = (component) => {
+    result = categoryValues.includes(highlightedValue);
+  }
+
+  // user is actively filtering; open all categories and show suggestions
```
```suggestion
  // user is actively filtering; default all categories to open
```

kyleschustak commented:
Pulled down this branch and did some testing. Here are some thoughts.

- Categories aren't highlighting on hover.
- When I expand a category it's changing from being below the box to above:
![image](https://github.com/narmi/design_system/assets/51931432/2fdc95c3-b704-4e3d-82eb-43072c1364b9)
![image](https://github.com/narmi/design_system/assets/51931432/3dab6618-d246-4c1b-8c67-316d25766420)
- When I select something, click off, then click back on, it treats it like a text input and nothing opens up
- I can just type some random text in the box and it seems like it's valid:
![image](https://github.com/narmi/design_system/assets/51931432/e13c683e-38cd-4e80-b670-84c6ebd2dce1)
- It's searching just by the beginning of the string, so we can't search by name of the category _and_ number for the category, it's one or the other

akdetrick commented:
Added some changes to address some of these issues.

> - Categories aren't highlighting on hover.

Categories now change background on CSS mouse hover. Categories will still be skipped when navigating via keyboard.

---

> - When I expand a category it's changing from being below the box to above:

Set a preferred side for top/bottom placement in the `useLayer` hook. This is otherwise expected behavior if there's not enough room on the page, but the switch to top positioning should happen less often.

❓ Is this considered a blocker? If so, I could add `onOpen` and `onClose` callbacks you could use to add some extra space on a parent element somewhere to guarantee there's room below the input.

---

> - When I select something, click off, then click back on, it treats it like a text input and nothing opens up

**Fix in progress** Will update this comment when I push it up

---

> - I can just type some random text in the box and it seems like it's valid:

Validation is a job for the parent component, and should probably be handled by showing an error that the value isn't valid. We don't want to reject input from the component as it would be a poor UX (too short of a timeout to reject means the input clears while the user is typing/correcting; too long a timeout means that the user thinks they've completed the field for the value to disappear if it's invalid)

---

> - It's searching just by the beginning of the string, so we can't search by name of the category _and_ number for the category, it's one or the other

Fixed via the new `filterItemsByInput` prop that takes a function
<img width="879" alt="Screenshot 2023-09-20 at 4 00 51 PM" src="https://github.com/narmi/design_system/assets/231252/afb6c696-8692-430a-90f6-7789f2ce8c9a">

kyleschustak commented on src/Combobox/index.stories.js:
```diff
@@ -150,6 +150,82 @@ export const InADialog = () => {
+      </Combobox.Item>
+    </Combobox.Category>,
+    <Combobox.Category label="Savings">
+      <Combobox.Item searchValue="Business Checking" value="savings1">
```
Minor, but think there's some copy/pasta here. Think you meant "Business Savings"
